tests/tensor/rewriting/test_elemwise.py

```diff
@@ -900,33 +928,120 @@ def my_init(dtype="float64", num=0):
                 fxv * np.sin(fsv),
                 "float32",
             ),
+            # Multiple output cases  # 72
+            (
+                (
+                    # sum(logp)
+                    at_sum(-((fx - fy) ** 2) / 2),
+                    # grad(logp)
+                    at.grad(at_sum(-((fx - fy) ** 2) / 2), wrt=fx),
+                ),
+                (fx, fy),
+                (fxv, fyv),
+                3,
+                (
+                    np.sum(-((fxv - fyv) ** 2) / 2),
+                    -(fxv - fyv),
+                ),
+                ("float32", "float32"),
+            ),
+            # Two Composite graphs that share the same input, but are split by
+            # a non-elemwise operation (Assert)
+            (
+                (
+                    log(
+                        ge(
+                            assert_op(
+                                at_abs(fx),
+                                at_all(
+                                    ge(
+                                        at_abs(fx),
+                                        0,  # Shared input
+                                    )
+                                ),
+                            ),
+                            0,
+                        )
+                    ),
+                ),
+                (fx,),
+                (fxv,),
+                4,
+                (np.zeros_like(fxv),),
+                ("float32",),
+            ),
+            # Two subgraphs that share the same non-fuseable input, but are otherwise
+            # completely independent
+            (
+                (
+                    true_div(
+                        mul(
+                            at_sum(fx + 5),  # breaks fusion
+                            exp(fx),
+                        ),
+                        (fx + 5),
+                    ),
+                ),
+                (fx,),
+                (fxv,),
+                4,
+                (np.sum(fxv + 5) * np.exp(fxv) / (fxv + 5),),
+                ("float32",),
+            ),
+            pytest.param(
+                (
+                    (sin(exp(fx)), exp(sin(fx))),
+                    (fx,),
+                    (fxv,),
+                    1,
+                    (np.sin(np.exp(fxv)), np.exp(np.sin(fxv))),
+                    ("float32", "float32"),
+                ),
+                marks=pytest.mark.xfail,  # Not implemented yet
+            ),
```

Comment:
I don't want to sound nitpicky, but you're adding more lines to the `parametrize` (74) than the test method has to begin with (61).

I mean sure, there were 600 lines of `parameterize` before, but like... _maybe_ some of these `parametrize` items deserve their own test method

Reply:
This is a rewrite that affects 99% of all graphs in a very fundamental manner, I don't think it's crazy that we would test so many parametrizations.

Reply:
I fully support having all these tests. I'm just questioning whether this is the most maintainable way to organize the test code.
For example, the test implementation could be extracted into a function, and some of the most involved parametrizations that have these 10ish levels of indentation could just be their own test methods.

Reply:
Hmm... We can also use a generator to yield the parametrizations, but I am not sure that's much better? I went through the parametrizations and there is only one that seems too long. I rearranged it to have less nesting:

```python
            (
                (
                    log(
                        ge(
                            assert_op(
                                at_abs(fx),
                                at_all(ge(at_abs(fx), 0)),
                            ),
                            0,
                        )
                    ),
                ),
                (fx,),
                (fxv,),
                4,
                (np.zeros_like(fxv),),
                ("float32",),
            ),
```